Fill out more scheme builtins

dolio opened the pull request:
This PR includes various builtin additions, mainly for scheme, the most significant of which are sandboxing related.

- `Value.dependencies` and `Code.dependencies` will be included, as they are auto-generated from unison, and will appear when the share version is incremented
- `Code.isMissing` has been implemented in scheme
- A system mirroring the sandbox checks has been implemented in scheme. This keeps track of sensitive builtins for each defined function, and looks them up to check if a value's evaluation should be allowed
- A couple new sandbox related builtins have been added:
    * `sandboxLinks` looks up the list of sensitive builtins associated with a termlink. This is used when statically generating code, to just look up the sandboxing information in the current runtime and add it to the file.
    * `Value.validateSandboxed` is a revised version of the sandboxing check. It allows checking the sandboxing status of a `Value` *before* it is loaded into an arbitrary unison value. This is a bit easier to check on the scheme side. It also has a different API, which can report all the builtins that caused the value to fail the sandbox check, rather than just a boolean.

Some tests are included, but I'm going to add a bit more, as I think there might be at least one corner case I missed on the scheme end.

dolio committed:
Add some new builtins for accessing sandbox information

- `Value.validateSandboxed` checks an interchange value for sandboxing
  before it is decoded. It either returns a list of dependencies the
  runtime doesn't know about yet, or a list of disallowed builtins, both
  as lists of termlinks
- `sandboxLinks` returns the list of sandboxed builtins transitively
  referenced by the code associated to a given termlink

dolio committed:
Implement sandboxing functions in scheme

- Creates a map storing associations of termlinks with sensitive ops
- Procedures for dynamically expanding the map, and declaring
  associations statically
- Structural termlink equality and hashing has been added to the struct,
  because otherwise looking up associated links was failing.
- Declarations for all the sensitive builtins have been included
  statically in the sandbox implementation file
  * Generated code will use a direct declaration of all builtins
    transitively used, since we can look that up at code generation time
- Both sandbox validation functions have been provided, but the one for
  arbitrary values is incomplete.
- Sandboxing has been added to the jit test suite.

dolio committed:
Update transcripts for new builtins

dolio committed:
Fix some type mismatches in primops

Lists vs. chunked lists, byte strings vs. typelinks, etc.

dolio committed:
Fix interpreter tests

- Value.validateSandboxed wasn't being aliased there
- The sandbox check in the interpreter wasn't handling builtin
  dependencies correctly

dolio committed:
Add alias for Value.validateSandboxed to jit tests

dolio committed:
Implement isMissing in scheme

dolio committed:
Put hard sandbox checks on new primops

pchiusano commented on parser-typechecker/src/Unison/Runtime/Machine.hs:
```diff
@@ -388,17 +393,36 @@ exec !env !denv !_activeThreads !ustk !bstk !k _ (BPrim1 DBTX i)
-  pure (denv, ustk, bstk, k)
+exec !env !denv !_activeThreads !ustk !bstk !k _ (BPrim2 SDBX i j)
+  | sandboxed env =
+      die "attempted to use sandboxed operation: sandboxLinks"
```
Should the error message say "sandboxLinks" here or something else?

dolio replied:
The error message is correct, but I'd put it in the wrong case there. It was meant for `SDBL`.

dolio committed:
Code formatting

pchiusano commented on scheme-libs/racket/unison/sandbox.rkt:
```diff
@@ -0,0 +1,149 @@
+(sandbox-builtin "IO.array")
+(sandbox-builtin "IO.arrayOf")
+(sandbox-builtin "IO.bytearray")
+(sandbox-builtin "IO.bytearrayOf")
```
I went ahead and double checked these to make sure they're consistent with what's in `Builtin.hs`.

pchiusano reviewed (approved):
LGTM, left one very minor comment. I'd say merge whenever you're ready.

dolio committed:
Accidentally put sandbox check on SDBX instead of SDBL

dolio committed:
Add a new serialized test case

Contains a serialized, mutually recursive function that won't be
recognized as unsafe unless the sandboxing information is expanded
correctly.

dolio committed:
Fix CACH implementation in scheme

It was failing to check the case where we already have all dependencies,
which was causing a crash when generating the module name, since we use
the termlinks to influence the name.

dolio committed:
Fix unsafe deserial test case check

Was wrapping in value too many times, which makes things safe.

dolio committed:
Corrected dynamic sandboxing algorithm

CACH wasn't actually adding to the sandbox, and the sandbox expansion
algorithm wasn't correctly handling mutually recursive functions.

dolio committed:
Tweak sandbox tests

Add some cases that actually fail if statically generated sandboxing
information isn't produced.

dolio committed:
Bump default JIT version
